File: packages/engine-community/android/src/main/kotlin/com/guardian/engine/community/HookDetector.kt
package com.guardian.engine.community

import android.content.pm.PackageManager

object HookDetector {

    private val KNOWN_HOOK_PACKAGES = listOf(
        "de.robv.android.xposed.installer",
        "com.saurik.substrate",
        "com.topjohnwu.magisk",
        "io.github.lsposed.manager",
        "org.lsposed.manager",
    )

    data class Result(
        val detected: Boolean,
        val confidence: Float,
        val evidence: Map<String, String>,
    )

    fun detect(packageManager: PackageManager): Result {
        val found = KNOWN_HOOK_PACKAGES.filter { pkg ->
            try {
                @Suppress("DEPRECATION")
                packageManager.getPackageInfo(pkg, 0)
                true
            } catch (_: PackageManager.NameNotFoundException) {
                false
            }
        }

        return Result(
            detected = found.isNotEmpty(),
            confidence = if (found.isNotEmpty()) 0.99f else 0.03f,
            evidence = if (found.isNotEmpty()) mapOf("packages" to found.joinToString(",")) else emptyMap(),
        )
    }
}

File: packages/engine-community/src/detectors/HookDetector.ts
import type { Detector, DetectorResult } from './Detector.js';

const KNOWN_HOOK_PACKAGES = ['de.robv.android.xposed.installer', 'com.saurik.substrate'] as const;

export class HookDetector implements Detector {
  readonly threatId = 'hooks' as const;
  readonly severity = 'critical' as const;

  async run(): Promise<DetectorResult> {
    const simulateHooks = process.env['GUARDIAN_SIMULATE_HOOKS'] === '1';
    if (!simulateHooks) {
      return { detected: false, confidence: 0.05, evidence: { method: 'js-stub' } };
    }
    return {
      detected: true,
      confidence: 0.99,
      evidence: {
        method: 'simulated',
        packages: KNOWN_HOOK_PACKAGES.join(','),
      },
    };
  }
}
